### PR TITLE
Student UI: Improve lecture series info

### PR DESCRIPTION
**Pull request by Coenego**
I know the original design specified a tooltip, but now having actually used the info icon / info popup in the student UI, I think the comprehensive info should be lifted into a modal. This approach would also give a better handling of OT events than the column based approach.

Items:
* [x] Lift content into a modal. Invoke the modal when the info icon in the list is clicked or tapped.
* [x] Add scrollable event list with date info, in date order which would take care of OT event handling
* [x] Style dates as we do in the admin UI 
* [x] Add minimal series type info for each event, with tooltip explanation on hover (just basic)
* [x] Include tripos and part in the modal title
* [x] Concatenate all lecturers and venues at the bottom of the modal

adj-student-series-info

### CODE REVIEW

**Comment by Coenego** [comment]
Assigning to @bp323 for review

File: shared/gh/css/gh.components.css [hunk 1]
```diff
@@ -1211,3 +1188,78 @@ span.gh-agenda-view-middot {
...
+
+/*********************/
+/* SERIES INFO MODAL */
+/*********************/
+
+#gh-series-info-modal .modal-header {
+    padding: 25px 25px 0 !important;
+}
+
+
```
Comment (bp323) [inline shared/gh/css/gh.components.css]: Extra line can go

File: shared/gh/js/utils/gh.utils.templates.js [hunk 2]
```diff
@@ -139,4 +139,40 @@ define(['exports', 'gh.constants'], function(exports, constants) {
...
     };
+
+    /**
+     * Render the hierarchy string
+     *
+     * @param  {Object}    orgUnit      The organisational unit to start building the hierarchy structure with
+     * @param  {String}    separator    The string used to split the organisational units
+     * @return {String}                 The generated hierarchy string
+     */
+    var renderHierarchyString = exports.renderHierarchyString = function(orgUnit, separator) {
```
Comment (bp323) [inline shared/gh/js/utils/gh.utils.templates.js]: Will we be using this in other areas as well? If not, I'd move this over to the modal.

File: shared/gh/js/utils/gh.utils.templates.js [hunk 3]
```diff
@@ -139,4 +139,40 @@ define(['exports', 'gh.constants'], function(exports, constants) {
...
+        }
+
+        // Store the organisational units' display names
+        var hierarchy = [];
+
+        /**
+         * Retrieve the display name for each parent object in the tree structure
+         *
+         * @param  {Object}    The organisational unit to return the display name from
+         * @private
```
Comment (bp323) [inline shared/gh/js/utils/gh.utils.templates.js]: Missing return here

File: shared/gh/js/utils/gh.utils.templates.js [hunk 4]
```diff
@@ -139,4 +139,40 @@ define(['exports', 'gh.constants'], function(exports, constants) {
...
+            throw new Error('An invalid value for orgUnit has been provided');
+        } else if (!_.isString(separator)) {
+            throw new Error('An invalid value for separator has been provided');
+        }
+
+        // Store the organisational units' display names
+        var hierarchy = [];
+
+        /**
+         * Retrieve the display name for each parent object in the tree structure
```
Comment (bp323) [inline shared/gh/js/utils/gh.utils.templates.js]: missing the bit about returning a string

File: shared/gh/js/views/gh.series-info.js [hunk 5]
```diff
@@ -159,58 +155,56 @@ define(['gh.core', 'gh.constants', 'moment', 'clickover'], function(gh, constant
...
+                'seriesTitle': seriesTitle
             }
-        };
+        }, $('#gh-modal'));
 
-        $trigger.clickover(options);
-        $trigger.trigger('click');
-    };
+        // Retrieve the series when the modal is shown
+        $('#gh-series-info-modal').on('shown.bs.modal', function(e) {
```
Comment (bp323) [inline shared/gh/js/views/gh.series-info.js]: Should this move to `addBinding`?

File: shared/gh/partials/series-info.html [hunk 6]
```diff
@@ -11,27 +9,42 @@ <h3 class="gh-popover-title"><%- data.displayName %></h3>
...
     <% } else { %>
-        <div class="col-sm-12">No events</div>
+        <div class="gh-series-info-list-no-events">No events</div>
     <% } %>
 </div>
 
+<!-- Organisers -->
 <% if (data.organisers && data.organisers.length) { %>
-    <p class="gh-popover-organiser"><%- data.organisers.join(', ') %></p>
+    <div class="gh-series-info-organiser"><%- data.organisers.join(', ') %></div>
```
Comment (bp323) [inline shared/gh/partials/series-info.html]: If there are no known organisers the copy should be `Lecturer not known`

File: shared/gh/partials/series-info.html [hunk 7]
```diff
@@ -11,27 +9,42 @@ <h3 class="gh-popover-title"><%- data.displayName %></h3>
...
 <% } %>
 
+<!-- Locations -->
 <% if (data.locations && data.locations.length) { %>
-    <p class="gh-popover-location"><%- data.locations.join(', ') %></p>
+    <div class="gh-series-info-location"><%- data.locations.join(', ') %></div>
```
Comment (bp323) [inline shared/gh/partials/series-info.html]: If there is no known location the copy should be `Location not known`

**Comment by bp323** [comment]
I'm seeing horizontal scrollbars in the modal content

screen shot 2015-04-15 at 16 41 05

File: shared/gh/css/gh.components.css [hunk 8]
```diff
@@ -1211,3 +1188,78 @@ span.gh-agenda-view-middot {
...
+    padding-top: 20px;
+}
+
+#gh-series-info-modal .gh-series-info-list .gh-series-info-list-no-events {
+    padding: 10px;
+}
+
+#gh-series-info-modal .gh-series-info-list .gh-series-info-list-term strong {
+    border-bottom-style: solid;
+    border-bottom-width: 1px;
```
Comment (bp323) [inline shared/gh/css/gh.components.css]: I -think- these are `2px` borders, looking at the design

Comment (bp323) [inline shared/gh/css/gh.components.css]: Feel free to ignore this if it doesn't look as good

**Comment by bp323** [comment]
Can we give the modal body a bit more height to work with? It's very cramped and the design also seems to be slightly higher than the implementation.

**Comment by bp323** [comment]
For longer hierarchies, the string overlaps with the `x` in the top right corner.

screen shot 2015-04-15 at 16 44 08

File: shared/gh/js/views/gh.series-info.js [hunk 9]
```diff
@@ -159,58 +155,56 @@ define(['gh.core', 'gh.constants', 'moment', 'clickover'], function(gh, constant
...
+        });
 
-        var options = {
-            'class_name': 'gh-series-popover gh-series-info-popover',
-            'container': 'body',
-            'content': $content.html(),
-            'global_close': true,
-            'html': true,
+        // Generate the hierarchy string
+        var hierarchy = gh.utils.renderHierarchyString(parent, ' > ');
```
Comment (bp323) [inline shared/gh/js/views/gh.series-info.js]: `fa-angle-double-right` should be used per design

Comment (bp323) [inline shared/gh/js/views/gh.series-info.js]: Although actually that's not consistent with the designs for agenda view. I'll leave it up to you to decide which looks better.

**Comment by bp323** [comment]
I ran into an error opening the modal for a series

screen shot 2015-04-15 at 16 48 23
screen shot 2015-04-15 at 16 48 39

**Comment by bp323** [comment]
When no events where added yet in a series, let's do the same thing as batch edit terms and my agenda weeks do (the grey bar with a descriptive message in it)

screen shot 2015-04-15 at 16 49 45

**Comment by bp323** [comment]
Review done, left a few code and functional comments. Back to @Coenego

File: shared/gh/js/utils/gh.utils.templates.js (hunk 2 above)
Comment (Coenego) [inline shared/gh/js/utils/gh.utils.templates.js]: This can be reused where we need to display the hierarchy.

**Comment by Coenego** [comment]
Reassigning to @bp323 for review after follow-up

**Comment by bp323** [comment]
It looks like ampersands are incorrectly escaped in the modal title

**Comment by bp323** [comment]
If there are no events the 'no lecturer' and 'no location' message should be hidden.

**Comment by bp323** [comment]
Maybe we could add the title next to the dates as well?

**Comment by bp323** [comment]
Looks good, back to @Coenego for a few follow up items

**Comment by Coenego** [comment]
Back to @bp323 for review after follow-up

**Comment by bp323** [comment]
Verified and merged.

**Pull request merged**